File: src/PidController.cpp
////////////////////////////////////////////////////////////////////////////
//
// PidController
//
// Class implement a PID loop
//
////////////////////////////////////////////////////////////////////////////

#include <cmath>
#include <limits>
#include <ros/ros.h>
#include <boost/algorithm/clamp.hpp>
#include "iarc7_msgs/Float64ArrayStamped.h"
#include "iarc7_motion/PidController.hpp"

using namespace Iarc7Motion;

PidController::PidController(double settings[6],
                             std::string debug_pid_name,
                             ros::NodeHandle& nh)
    : p_gain_(settings[0]),
      i_gain_(settings[1]),
      d_gain_(settings[2]),
      initialized_(false),
      i_accumulator_(0.0),
      last_current_value_(0.0),
      last_time_(0.0),
      setpoint_(0.0),
      i_accumulator_max_(settings[3]),
      i_accumulator_min_(settings[4]),
      i_accumulator_enable_threshold_(settings[5])
{
    pid_value_publisher_ = nh.advertise<iarc7_msgs::Float64ArrayStamped>(debug_pid_name, 1000);
}

bool PidController::update(double current_value,
                           const ros::Time& time,
                           double& response,
                           double derivative,
                           bool log_debug)
{
    if (time < last_time_) {
        ROS_WARN("Time passed in to PidController is less than the last time.");
        return false;
    }

    if(time == last_time_)
    {
        ROS_WARN("Time passed in to PidController is equal to the last time.");
        return false;
    }

    if (!std::isfinite(current_value)) {
        ROS_WARN("Invalid argument to PidController::update (current_value = %f)",
                 current_value);
        return false;
    }

    if (!std::isnan(derivative) && !std::isfinite(derivative)) {
        ROS_WARN("Invalid argument to PidController::update (derivative = %f)",
                 derivative);
        return false;
    }

    double difference = setpoint_ - current_value;
    double p_term = p_gain_ * difference;
    response = p_term;

    if (!initialized_) {
        initialized_ = true;
    } else {
        double time_delta = (time - last_time_).toSec();

        if (std::abs(difference) < i_accumulator_enable_threshold_) {
            i_accumulator_ += i_gain_ * difference * time_delta;

            boost::algorithm::clamp(i_accumulator_, i_accumulator_min_, i_accumulator_max_);
        } 

        response += i_accumulator_;

        // Check if we were passed a derivative or not
        if (std::isnan(derivative)) {
            derivative = (current_value - last_current_value_) / time_delta;
        }

        double d_term = d_gain_ * derivative;
        response -= d_term;

        //Publish PID values to topic
        if (log_debug) {
            iarc7_msgs::Float64ArrayStamped debug_msg;
            debug_msg.header.stamp = time;
            debug_msg.data = {p_term, i_accumulator_,-d_term};
            pid_value_publisher_.publish(debug_msg);
        }
    }

    last_current_value_ = current_value;
    last_time_ = time;

    if (!std::isfinite(response)) {
        ROS_WARN("Invalid result from PidController::update (response = %f)",
                 response);
        return false;
    } else {
        return true;
    }
}

void PidController::setSetpoint(double setpoint)
{
    setpoint_ = setpoint;
}

void PidController::resetAccumulator()
{
    i_accumulator_ = 0;
}

File: src/iarc7_motion/iarc_tasks/task_utilities/pid_controller.py
#!/usr/bin/env python
import rospy
import numpy as np

class PidSettings(object):
    def __init__(self, settings_dict):
        self.kp = settings_dict['kp']
        self.ki = settings_dict['ki']
        self.kd = settings_dict['kd']
        self.accum_max = settings_dict['accumulator_max']
        self.accum_min = settings_dict['accumulator_min']
        self.accum_en_threshold = settings_dict['accumulator_enable_threshold']
        self.i_accumulator_initial_value = None

class PidController(object):
    def __init__(self, settings):
        self._p_gain = settings.kp
        self._i_gain = settings.ki
        self._d_gain = settings.kd
        self._i_accumulator_max = settings.accum_max
        self._i_accumulator_min = settings.accum_min
        self._i_accumulator_enable_threshold = settings.accum_en_threshold

        self._initialized = False
        self._last_time = None

        if settings.i_accumulator_initial_value is not None:
            self._i_accumulator = settings.i_accumulator_initial_value
        else:
            self._i_accumulator = 0.0

        self._last_current_value = 0.0
        self._setpoint = 0.0

    def reset_accumulator(self):
        self._i_accumulator = 0.0

    def set_setpoint(self, setpoint):
        self._setpoint = setpoint

    def get_accumulator(self):
        return self._i_accumulator

    def update(self, current_value, time, log_debug):
        '''
        Updates PID controller

        Args:
            current_value: current value of whatever you are controlling on
            time: current time; use rospy.get_time()
            log_debug: enables verbose debugging

        Returns:
            success: whether or not a response could be calculated
            response is the filter output
        '''
        response = 0

        if self._last_time is None:
            self._last_time = time
            self._initialized = True
            return True, response

        if time == self._last_time:
            rospy.logwarn(
                    'Time passed in to PidController is equal to last time.')
            return False, response

        if time < self._last_time:
            rospy.logwarn(
                    'Time passed in to PidController is less than the last time.')
            return False, response

        if not np.isfinite(current_value):
            rospy.logwarn(
                    'Invalid argument to PidControllor.update (current_value = {}).'.format(current_value))
            return False, response

        difference = self._setpoint - current_value
        p_term = self._p_gain * difference
        response = p_term

        time_delta = (time - self._last_time).to_sec()

        if np.absolute(difference) < self._i_accumulator_enable_threshold:
            self._i_accumulator = (self._i_accumulator
                                + self._i_gain * difference * time_delta)

            if log_debug:
                log = ('Set accumulator to ' + str(self._i_accumulator) +
                        ', i_gain to ' + str(self._i_gain) +
                        ', difference is ' + str(difference) +
                        ', time delta is ' + str(time_delta))
                rospy.logwarn(log)

            self._i_accumulator = max(min(self._i_accumulator_max,
                                          self._i_accumulator),
                                          self._i_accumulator_min)

            if log_debug:
                log = ('Accumulator clamped to ' + str(self._i_accumulator) +
                        ' (min: ' + str(self._i_accumulator_min) +
                        ', max: ' + str(self._i_accumulator_max) + ')')
                rospy.logwarn(log)
        else:
            if log_debug:
                log = ('Ignoring difference ' + str(difference) +
                        ' above threshold ' + str(self._i_accumulator_enable_threshold))
                rospy.logwarn(log)

        response = response + self._i_accumulator

        derivative = (current_value - self._last_current_value) / time_delta

        d_term = self._d_gain * derivative
        response = response - d_term

        if log_debug:
            log = ('p: ' + str(p_term) +
                    ' I: ' + str(self._i_accumulator) +
                    ' D: ' + str(d_term))

            rospy.logwarn(log)

        self._last_current_value = current_value
        self._last_time = time

        if not np.isfinite(response):
            rospy.logwarn('Invalid result from PidController.update (response = {})'.format(response))
            return False, response
        else:
            return True, response
